File: solutions/go/matching-brackets/matching_brackets.go
// Package brackets checks for pairs brackets.
package brackets

// Bracket checks for balanced brackets.
func Bracket(s string) bool {
	var stack []rune
	var r rune

	for _, c := range s {
		switch c {
		case '[':
			fallthrough
		case '(':
			fallthrough
		case '{':
			stack = append(stack, c)
			continue
		case ']':
			r = '['
		case ')':
			r = '('
		case '}':
			r = '{'
		default:
			continue
		}

		if len(stack) == 0 || stack[len(stack)-1] != r {
			return false
		}
		stack = stack[:len(stack)-1]
	}
	return len(stack) == 0
}

File: solutions/ruby/matching-brackets/matching_brackets.rb
module Brackets
  PAIRS = '{}()[]<>'.chars.each_slice(2).to_h
  OPEN = PAIRS.keys
  CLOSE = PAIRS.values

  def self.paired?(data)
    stack = []
    data.chars.each do |char|
      if OPEN.include?(char)
        stack.append(char)
      elsif CLOSE.include?(char)
        return false unless PAIRS[stack.pop] == char
      end
    end
    stack.empty?
  end
end
